### PR TITLE
Move to errgo

### PR DESCRIPTION
**Pull request by makyo**

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/bundleservice/13/
Test PASSed.

**Comment by mhilton** [comment]
:+1:

File: params/params.go [hunk 1]
```diff
@@ -4,6 +4,19 @@ package params
 
 import "github.com/juju/httprequest"
 
+// ErrorCode allows for const errors
```
Comment (rogpeppe) [inline params/params.go]: // ErrorCode holds the code of an error returned from the API.

?

File: params/params.go [hunk 2]
```diff
@@ -4,6 +4,19 @@ package params
...
+type ErrorCode string
+
+// Error returns the code as a string
+func (code ErrorCode) Error() string {
```
Comment (rogpeppe) [inline params/params.go:21]: // Error implements error.Error.

? (it's a standard interface in Go)

File: params/params.go [hunk 3]
```diff
@@ -4,6 +4,19 @@ package params
...
+	return string(code)
+}
+
+// Error constants provided to be used by the server.
```
Comment (rogpeppe) [inline params/params.go]: // Possible error codes from the API.

? (the client can use 'em too)

File: server.go [hunk 4]
```diff
@@ -41,7 +41,13 @@ type errorResponse struct {
 // errorMapper maps an error from a handler into an HTTP server error.
 // TODO map other errors
 var errorMapper httprequest.ErrorMapper = func(err error) (int, interface{}) {
-	return http.StatusInternalServerError, &errorResponse{
+	status := http.StatusInternalServerError
+	cause := errgo.Cause(err)
+	switch cause {
+	case params.ErrUnparsable:
+		status = 422 // "Unprocessable Entity" - http doesn't have a const for this error.
```
Comment (rogpeppe) [inline server.go]: Ha, that's a new one on me. Perhaps define a constant above:

```
const StatusUnprocessableEntity = 422
```

Or just return StatusBadRequest, I guess.

File: server.go [hunk 5]
```diff
@@ -41,7 +41,13 @@ type errorResponse struct {
...
+	}
```
Comment (rogpeppe) [inline server.go:45]: I think you probably want to return Code: params.ErrUnparsable in this case.

Comment (rogpeppe) [inline server.go:45]: Perhaps something like:

```
 code, _ := cause.(params.ErrorCode)
 return status, &errorResponse{
      Message: err.Error(),
      Code: code,
 }
```

File: server.go [hunk 6]
```diff
@@ -64,11 +69,11 @@ func (h *handler) GetChangesFromYAML(p *params.ChangesFromYAMLParams) (params.Ch
 func getChanges(bundleYAML string) ([]params.Change, error) {
 	bundle, err := charm.ReadBundleData(strings.NewReader(bundleYAML))
 	if err != nil {
-		return nil, fmt.Errorf("error reading bundle data: %v", err)
+		return nil, errgo.WithCausef(err, params.ErrUnparsable, "error reading bundle data")
 	}
 	err = bundle.Verify(nil, nil)
 	if err != nil {
-		return nil, fmt.Errorf("error verifying bundle data: %v", err)
+		return nil, errgo.WithCausef(err, params.ErrUnparsable, "error verifying bundle data")
```
Comment (rogpeppe) [inline server.go]: Is "unparsable" really the right code here? I'm wondering if you want an ErrVerificationFailure as well.

File: server.go [hunk 7]
```diff
@@ -41,7 +41,13 @@ type errorResponse struct {
...
 // TODO map other errors
 var errorMapper httprequest.ErrorMapper = func(err error) (int, interface{}) {
-	return http.StatusInternalServerError, &errorResponse{
+	status := http.StatusInternalServerError
+	cause := errgo.Cause(err)
+	switch cause {
+	case params.ErrUnparsable:
+		status = 422 // "Unprocessable Entity" - http doesn't have a const for this error.
+	}
+	return status, &errorResponse{
```
Comment (rogpeppe) [inline server.go]: Probably for a subsequent PR but you probably want to define errorResponse publicly in params. That way you can check for it specifically in this mapper, and that gives API endpoints the freedom to return custom errors directly (you could add an "Info" field to hold all the individual verification errors, for example).

**Comment by rogpeppe** [comment]
LGTM with a bunch of mostly-trivials. Thanks!

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/bundleservice/14/
Test PASSed.

**Comment by makyo** [comment]
Thanks all :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/bundleservice-merge

**Pull request merged**